### PR TITLE
[cpu] Vectorize nan_to_num

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #98329

Locally I see a roughly 4x speedup.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

### CODE REVIEW

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 1]
```diff
@@ -444,57 +444,76 @@ static void polygamma_kernel(TensorIteratorBase& iter, int64_t n) {
...
+inline Vectorized<c10::complex<scalar_t>> _nan_to_num_replace(
+    const Vectorized<c10::complex<scalar_t>> &a, scalar_t nan, scalar_t posinf, scalar_t neginf) {
+#if !defined(_MSC_VER) && (defined(CPU_CAPABILITY_AVX2) || defined(CPU_CAPABILITY_AVX512))
+  return {_nan_to_num_replace(Vectorized<scalar_t>(a), nan, posinf, neginf)};
+#else
+  __at_align__ c10::complex<scalar_t> buffer[a.size()];
+  a.store(buffer);
+  auto asreal = Vectorized<scalar_t>::loadu(buffer);
+  _nan_to_num_replace(asreal, nan, posinf, neginf).store(buffer);
+  return Vectorized<c10::complex<scalar_t>>::loadu(buffer);
```
Comment (lezcano) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:491]: Why is MSVC complaining here?

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 2]
```diff
@@ -444,57 +444,76 @@ static void polygamma_kernel(TensorIteratorBase& iter, int64_t n) {
...
+  vec_t inf(std::numeric_limits<scalar_t>::infinity());
+  vec_t result;
+  result = vec_t::blendv(a, vec_t(nan), a.isnan());
+  result = vec_t::blendv(result, vec_t(posinf), a == inf);
+  return vec_t::blendv(result, vec_t(neginf), a == inf.neg());
+}
+
+template <typename scalar_t>
+inline Vectorized<c10::complex<scalar_t>> _nan_to_num_replace(
+    const Vectorized<c10::complex<scalar_t>> &a, scalar_t nan, scalar_t posinf, scalar_t neginf) {
```
Comment (lezcano) [inline aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:483]: Nit. Vectorized should be passed by copy, as they fit in a register.

**Review by lezcano — APPROVED** [review]
Cool, thanks!

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp (hunk 1 above)
Comment (peterbell10) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:491]: The code uses a conversion via vector intrinsics like `__m256` but `Vectorized` isn't specialized on MSVC so these conversion operators don't exist:
https://github.com/pytorch/pytorch/blob/54c28c564f0f4da174358312a026398ebca5337c/aten/src/ATen/cpu/vec/vec256/vec256_float.h#L18

**Pull request closed without merging**